transports/webtransport-websys/src/stream.rs

```diff
@@ -0,0 +1,245 @@
+                .maybe_init_and_poll(cx, || self.writer.ready()))
+            .map_err(to_io_error)?;
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_writer_close(&mut self, cx: &mut Context) -> Poll<io::Result<()>> {
+        match self.writer_state {
+            StreamState::Open => {
+                self.writer_state = StreamState::Closing;
+
+                // Initiate close
+                detach_promise(self.writer.close());
+
+                // Assume closed on error
+                let _ = ready!(self
+                    .writer_closed_promise
+                    .maybe_init_and_poll(cx, || self.writer.closed()));
+
+                self.writer_state = StreamState::Closed;
+            }
+            StreamState::Closing => {
+                // Assume closed on error
+                let _ = ready!(self.writer_closed_promise.poll(cx));
+                self.writer_state = StreamState::Closed;
+            }
+            StreamState::Closed => {}
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_reader_read(&mut self, cx: &mut Context) -> Poll<io::Result<Option<Uint8Array>>> {
+        let val = ready!(self
+            .reader_read_promise
+            .maybe_init_and_poll(cx, || self.reader.read()))
+        .map_err(to_io_error)?;
+
+        let val = parse_reader_response(&val)
+            .map_err(to_io_error)?
+            .map(Uint8Array::from);
+
+        Poll::Ready(Ok(val))
+    }
+}
+
+impl Drop for Stream {
+    fn drop(&mut self) {
+        // Close writer.
+        //
+        // We choose to use `close()` instead of `abort()`, because
+        // abort was causing some side effects on the WebTransport
+        // layer and connection was lost.
+        detach_promise(self.writer.close());
+
+        // Cancel any ongoing reads.
+        detach_promise(self.reader.cancel());
+    }
+}
+
+impl AsyncRead for Stream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let this = self.get_mut();
+
+        // If we have leftovers from a previous read, then use them.
+        // Otherwise read new data.
+        let data = match this.read_leftovers.take() {
+            Some(data) => data,
+            None => {
+                match ready!(this.poll_reader_read(cx))? {
+                    Some(data) => data,
+                    // EOF
+                    None => return Poll::Ready(Ok(0)),
+                }
+            }
+        };
+
+        if data.byte_length() == 0 {
+            return Poll::Ready(Ok(0));
+        }
+
+        let out_len = data.byte_length().min(buf.len() as u32);
+        data.slice(0, out_len).copy_to(&mut buf[..out_len as usize]);
+
+        let leftovers = data.slice(out_len, data.byte_length());
+
+        if leftovers.byte_length() > 0 {
+            this.read_leftovers = Some(leftovers);
+        }
+
+        Poll::Ready(Ok(out_len as usize))
```

Comment:
Stuff like this could be unit-tested by introduced a dedicated type for `read_leftovers` and testing the interface of that.

Reply:
I choose to keep implement it like this for simplicity, it is still unit testable because of how QUIC works. What do you prefer:

1. A unit test that will cover `read_leftovers`
2. Dedicated types. In this case I will introduce `JsStream` and `JsBufReader`, and maybe `JsSink` for consistency.

Reply:
This was more of an idea for how we could regain some of the coverage if we remove the other tests. Let's work that out first :)